File: 2025/December/Day04/solution/main.py
import sys
import os
import time
from collections import deque


def main():
    total_start = time.perf_counter()

    file_path = sys.argv[1] if len(sys.argv) > 1 else "data/input.txt"

    if not os.path.exists(file_path):
        raise FileNotFoundError(f"Input file not found: {file_path}")

    # Read grid once
    with open(file_path, "r", encoding="utf-8") as f:
        grid = [list(line.rstrip("\n")) for line in f]

    rows = len(grid)
    columns = len(grid[0])

    # Directions for 8 neighbors (row, col pairs)
    directions = [-1, -1, -1, 0, -1, 1, 0, -1, 0, 1, 1, -1, 1, 0, 1, 1]

    # Part 1
    part1_start = time.perf_counter()
    count = 0

    for row in range(rows):
        for col in range(columns):
            if grid[row][col] != "@":
                continue

            neighbors = 0
            i = 0
            while i < len(directions):
                nr = row + directions[i]
                nc = col + directions[i + 1]

                if 0 <= nr < rows and 0 <= nc < columns and grid[nr][nc] == "@":
                    neighbors += 1
                i += 2

            if neighbors < 4:
                count += 1

    part1 = count
    part1_time = int((time.perf_counter() - part1_start) * 1000)

    # Part 2
    part2_start = time.perf_counter()

    # Copy of grid so it is editable
    mutable_grid = [row.copy() for row in grid]

    # Neighbor count grid
    neighbor_count = [[0] * columns for _ in range(rows)]

    def compute_neighbors(r, c):
        cnt = 0
        i = 0
        while i < len(directions):
            nr = r + directions[i]
            nc = c + directions[i + 1]
            if 0 <= nr < rows and 0 <= nc < columns and mutable_grid[nr][nc] == "@":
                cnt += 1
            i += 2
        return cnt

    # Initialize neighbor counts
    for r in range(rows):
        for c in range(columns):
            if mutable_grid[r][c] == "@":
                neighbor_count[r][c] = compute_neighbors(r, c)

    removed_total = 0
    queue = deque()

    # Store initially accessible paper rolls
    for r in range(rows):
        for c in range(columns):
            if mutable_grid[r][c] == "@" and neighbor_count[r][c] < 4:
                queue.append((r, c))

    # Iterative removal
    while queue:
        next_queue = deque()

        for r, c in queue:
            if mutable_grid[r][c] != "@":
                continue

            # Remove paper roll
            mutable_grid[r][c] = "."
            removed_total += 1

            # Update neighbors
            i = 0
            while i < len(directions):
                nr = r + directions[i]
                nc = c + directions[i + 1]

                if 0 <= nr < rows and 0 <= nc < columns and mutable_grid[nr][nc] == "@":
                    neighbor_count[nr][nc] -= 1

                    # If it just became accessible
                    if neighbor_count[nr][nc] == 3:
                        next_queue.append((nr, nc))
                i += 2

        queue = next_queue

    part2 = removed_total
    part2_time = int((time.perf_counter() - part2_start) * 1000)

    print("\n========= Answers =========")
    print(f"Part 1: {part1}")
    print(f"Part 2: {part2}")

    print("\n========= Timings =========")
    print(f"Part 1 Time: {part1_time}ms")
    print(f"Part 2 Time: {part2_time}ms")

    total_time = int((time.perf_counter() - total_start) * 1000)
    print(f"\nTotal Runtime: {total_time} ms\n")


if __name__ == "__main__":
    main()

File: 2025/December/Day04/solution/main.js
const fs = require("fs");
const path = require("path");

function main() {
  const totalStart = Date.now();

  const filePath = process.argv[2] || "data/input.txt";
  const resolvedPath = path.resolve(filePath);

  if (!fs.existsSync(resolvedPath)) {
    throw new Error(`Input file not found: ${filePath}`);
  }

  // Read grid once
  const grid = fs
    .readFileSync(resolvedPath, "utf8")
    .trimEnd()
    .split(/\r?\n/)
    .map((line) => line.split(""));

  const rows = grid.length;
  const columns = grid[0].length;

  // Directions for 8 neighbors (row, col pairs)
  const directions = [
    -1, -1, -1, 0, -1, 1,
    0, -1, 0, 1,
    1, -1, 1, 0, 1, 1
  ];

  let part1;
  let part2;

  // Part 1
  const part1Start = Date.now();
  let count = 0;

  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < columns; col++) {
      if (grid[row][col] !== "@") continue;

      let neighbors = 0;
      for (let i = 0; i < directions.length; i += 2) {
        const nr = row + directions[i];
        const nc = col + directions[i + 1];

        if (
          nr >= 0 && nr < rows &&
          nc >= 0 && nc < columns &&
          grid[nr][nc] === "@"
        ) {
          neighbors++;
        }
      }

      if (neighbors < 4) count++;
    }
  }

  part1 = count;
  const part1Time = Date.now() - part1Start;

  // Part 2
  const part2Start = Date.now();

  // Deep copy grid so it is editable
  const mutableGrid = grid.map((row) => row.slice());

  // Neighbor count grid
  const neighborCount = Array.from({ length: rows }, () =>
    Array(columns).fill(0)
  );

  function computeNeighbors(r, c) {
    let cnt = 0;
    for (let i = 0; i < directions.length; i += 2) {
      const nr = r + directions[i];
      const nc = c + directions[i + 1];
      if (
        nr >= 0 && nr < rows &&
        nc >= 0 && nc < columns &&
        mutableGrid[nr][nc] === "@"
      ) {
        cnt++;
      }
    }
    return cnt;
  }

  // Initialize neighbor counts
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < columns; c++) {
      if (mutableGrid[r][c] === "@") {
        neighborCount[r][c] = computeNeighbors(r, c);
      }
    }
  }

  let removedTotal = 0;
  let queue = [];

  // Store initially accessible paper rolls
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < columns; c++) {
      if (
        mutableGrid[r][c] === "@" &&
        neighborCount[r][c] < 4
      ) {
        queue.push([r, c]);
      }
    }
  }

  // Iterative removal
  while (queue.length > 0) {
    const nextQueue = [];

    for (const [r, c] of queue) {
      if (mutableGrid[r][c] !== "@") continue;

      // Remove paper roll
      mutableGrid[r][c] = ".";
      removedTotal++;

      // Update neighbors
      for (let i = 0; i < directions.length; i += 2) {
        const nr = r + directions[i];
        const nc = c + directions[i + 1];

        if (
          nr >= 0 && nr < rows &&
          nc >= 0 && nc < columns &&
          mutableGrid[nr][nc] === "@"
        ) {
          neighborCount[nr][nc]--;

          // If it just became accessible
          if (neighborCount[nr][nc] === 3) {
            nextQueue.push([nr, nc]);
          }
        }
      }
    }

    queue = nextQueue;
  }

  part2 = removedTotal;
  const part2Time = Date.now() - part2Start;

  console.log("\n========= Answers =========");
  console.log(`Part 1: ${part1}`);
  console.log(`Part 2: ${part2}`);

  console.log("\n========= Timings =========");
  console.log(`Part 1 Time: ${part1Time}ms`);
  console.log(`Part 2 Time: ${part2Time}ms`);

  const totalTime = Date.now() - totalStart;
  console.log(`\nTotal Runtime: ${totalTime} ms\n`);
}

main();
